Hooks FAQ (#116)
reactjs/pl.reactjs.org: content/docs/hooks-faq.md

  ).join('\n')
-->

> * **[Strategia adopcji](#adoption-strategy)**

Reviewer: Adoptuje się dziecko, kod czy rozwiązania się adaptuje 😉więc powinno być `Strategia adaptacji`

Author: Dałem `Strategia wdrażania` :)

  * [Które wersje Reacta wspierają hooki?](#which-versions-of-react-include-hooks)
  * [Czy muszę przepisać wszystkie komponenty klasowe?](#do-i-need-to-rewrite-all-my-class-components)
  * [Co mogę zrobić z hookami czego nie mogłem zrobić z klasami?](#what-can-i-do-with-hooks-that-i-couldnt-with-classes)
  * [Jaka część mojej wiedzy o Reakcie jest nadal aktualna?](#how-much-of-my-react-knowledge-stays-relevant)

>   * [Czego mam używać: hooków, klas, a może mieszać oba sposoby?](#should-i-use-hooks-classes-or-a-mix-of-both)

Reviewer: W moim odczuciu `czego` w tym kontekscie brzmi delikatnie mówiąc "szorstko". Może warto rozważyć coś w stylu: `Czy powinienem używać hooków, klas, a może mieszać obydwa sposoby?`

Reviewer 2: Dlaczego "czego" brzmi tutaj szorstko? Przecież to ma wydźwięku "Czego?!" jako agresywna odmiana "Słucham?" ;-)

>   * [Czy hooki pokrywają wszystkie przypadki użycia, które są dostępne dla klas?](#do-hooks-cover-all-use-cases-for-classes)

Reviewer: Może `pokrywają` zmienić na `obejmują`?

>   * [Czy hooki zastępują właściwości renderujące i komponenty wyższego rzędu?](#do-hooks-replace-render-props-and-higher-order-components)

Reviewer: Dodałbym cudzysłów, tak jak jest w dokumentacji: `"właściwości renderujące"`.

  * [Co hooki oznaczają dla popularnych API takich jak connect() z Reduxa lub React Router?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)
  * [Czy hooki współpracują ze statycznym typowaniem?](#do-hooks-work-with-static-typing)

>   * [Jak testować komponenty które używają hooków?](#how-to-test-components-that-use-hooks)

Reviewer 2: ```suggestion
  * [Jak testować komponenty, które używają hooków?](#how-to-test-components-that-use-hooks)
```

>   * [Czego dokładnie wymagają reguły lintera?](#what-exactly-do-the-lint-rules-enforce)

Reviewer: Skłaniałbym się ku `Co dokładnie narzucają reguły lintera?`

* **[Od klas do hooków](#from-classes-to-hooks)**
  * [Jak wyglądają metody cyklu życia w odniesieniu do hooków?](#how-do-lifecycle-methods-correspond-to-hooks)
  * [Jak mogę pobrać dane wykorzystując hooki?](#how-can-i-do-data-fetching-with-hooks)
  * [Czy istnieje coś podobnego do zmiennych instancji?](#is-there-something-like-instance-variables)

>   * [Lepiej używać jednego czy wielu zmiennych stanu?](#should-i-use-one-or-many-state-variables)

Reviewer: `jednej`

>   * [Powinienem używać jednego czy wielu zmiennych stanu?](#should-i-use-one-or-many-state-variables)

Reviewer 2: ```suggestion
  * [Lepiej używać jednego czy wielu zmiennych stanu?](#should-i-use-one-or-many-state-variables)
```

>   * [Czy mogę uruchomić efekt tylko podczas aktualizacji?](#can-i-run-an-effect-only-on-updates)

Reviewer: Coś nie do końca pasuje mi tutaj `aktualizacja`, ale samemu mi trudno znaleźć lepszy wyraz. Może po prostu zapisać to jako `Czy mogę uruchomić efekt tylko podczas nowych zmian?`

Reviewer: Co myślisz o: `Czy mogę pominąć efekt w przypadku nowych zmian?`

Reviewer 2: Jestem za dopisaniem "komponentu".

Author: Dzięki @idbartosz  i @jakubdrozdek . Wrzuciłem tłumaczenie z dopiskiem "komponentu" -> `Czy mogę pominąć efekt podczas aktualizacji komponentu?`. 👍

Reviewer 2: @bartlomiejzuber Nie widzę słowa "komponentu" - na pewno wrzuciłeś? ;)

  * [Jak dostać poprzednie propsy lub stan?](#how-to-get-the-previous-props-or-state)
  * [Dlaczego widzę nieaktualne propsy lub stan wewnątrz mojej funkcji?](#why-am-i-seeing-stale-props-or-state-inside-my-function)
  * [Jak zaimplementować getDerivedStateFromProps?](#how-do-i-implement-getderivedstatefromprops)
  * [Czy istnieje coś takiego jak forceUpdate?](#is-there-something-like-forceupdate)
  * [Czy mogę stworzyć referencję do komponentu funkcyjnego?](#can-i-make-a-ref-to-a-function-component)
  * [Jak mogę zmierzyć węzeł DOM?](#how-can-i-measure-a-dom-node)
  * [Co oznacza const [thing, setThing] = useState()?](#what-does-const-thing-setthing--usestate-mean)
* **[Optymalizacja wydajności](#performance-optimizations)**
  * [Czy mogę pominąć efekt podczas aktualizacji?](#can-i-skip-an-effect-on-updates)

>   * [Czy bezpieczne jest pomijać funkcję w listach zależności?](#is-it-safe-to-omit-functions-from-the-list-of-dependencies)

Reviewer: `funkcje` zamiast `funkcję`

>   * [Co mogę zrobić jeżeli zależności mojego efektu zmieniaja się zbyt często?](#what-can-i-do-if-my-effect-dependencies-change-too-often)

Reviewer 2: ```suggestion
  * [Co mogę zrobić, jeżeli zależności mojego efektu zmieniaja się zbyt często?](#what-can-i-do-if-my-effect-dependencies-change-too-often)
```

  * [Jak zaimplementować shouldComponentUpdate?](#how-do-i-implement-shouldcomponentupdate)
  * [Jak memoizować obliczenia?](#how-to-memoize-calculations)

>   * [Jak leniwie tworzyć ciężkie komponenty?](#how-to-create-expensive-objects-lazily)

Reviewer: Może, `Jak w leniwy sposób tworzyć "ciężkie" komponenty?`.

Reviewer 2: Zaraz, przecież tu nie jest mowa o komponentach, tylko o obiektach. Powinno być "Jak w leniwy sposób tworzyć "ciężkie" obiekty?".

Reviewer: Good catch @jakubdrozdek 👍

  * [Czy hooki są wolne z powodu tworzenia funkcji podczas renderowania?](#are-hooks-slow-because-of-creating-functions-in-render)

>   * [Jak unikać przekazywania pętli zwrotnych w dół?](#how-to-avoid-passing-callbacks-down)

Reviewer 2: ```suggestion
  * [Jak unikać przekazywania funkcji zwrotnych w dół?](#how-to-avoid-passing-callbacks-down)
```

>   * [Jak czytać często zmieniającą się wartość z useCallback?](#how-to-read-an-often-changing-value-from-usecallback)

Reviewer: `czytać` na `odczytywać`

* **[Pod maską](#under-the-hood)**
  * [Jak React łączy wywołania hooków z komponentami?](#how-does-react-associate-hook-calls-with-components)
  * [Jaki jest stan patentu dla hooków?](#what-is-the-prior-art-for-hooks)

> ## Strategia adpocji {#adoption-strategy}

Reviewer 2: ```suggestion
## Strategia wdrożenia {#adoption-strategy}
```

### Które wersje Reacta wspierają hooki? {#which-versions-of-react-include-hooks}

Zaczynając od wersji 16.8.0, React zawiera stabilną implementacje hooków dla:

* React DOM
* React DOM Server
* React Test Renderer
* React Shallow Renderer

> Zauważ, że **aby włączyć hooki, wszystkie paczki Reacta muszą mieć wersję 16.8.0 lub wyższą**. hooki nie zadziałają, jeżeli zapomnisz zaktualizować, na przykład React DOM.

Reviewer 3: ```suggestion
Zauważ, że **aby włączyć hooki, wszystkie paczki Reacta muszą mieć wersję 16.8.0 lub wyższą**. Hooki nie zadziałają, jeżeli zapomnisz zaktualizować, na przykład React DOM.
```

> Zauważ, że **aby włączyć hooki, wszystkie paczki Reacta muszą mieć wersję 16.8.0 lub wyższą**. Hooki nie zadziałają, jeżeli zapomnisz zaktualizować, na przykład React DOM.

Reviewer 2: ```suggestion
Zauważ, że **aby włączyć hooki, wszystkie paczki Reacta muszą mieć wersję 16.8.0 lub wyższą**. Hooki nie zadziałają, jeżeli zapomnisz zaktualizować, na przykład, React DOM.
```

> React Native w wersji 0.59 i wyższej również wspiera hooki.

Reviewer 3: ```suggestion
React Native wspiera hooki od wersji 0.59.
```

### Czy muszę przepisać wszystkie komponenty klasowe? {#do-i-need-to-rewrite-all-my-class-components}

> Nie. Usunięcia klas z Reacta [nie ma w planach](/docs/hooks-intro.html#gradual-adoption-strategy) -- my wszyscy, zawsze musimy dostarczać produkty i nie możemy sobie pozwolić na ich przepisywanie. Z tego powodu, zalecamy wypróbowanie hooków w nowym kodzie.

Reviewer 3: ```suggestion
Nie. [Nie ma planów](/docs/hooks-intro.html#gradual-adoption-strategy) aby usunąć klasy z Reacta -- wszyscy musimy stale dostarczać nasze produkty i nie możemy sobie pozwolić na ich przepisywanie. Zachęcamy do wypróbowania hooków w nowym kodzie.
```

> ### Co mogę zrobić z hookami czego nie mogłem zrobić z klasami? {#what-can-i-do-with-hooks-that-i-couldnt-with-classes}

Reviewer 3: ```suggestion
### Co mogę zrobić z hookami, czego nie mogłem zrobić z klasami? {#what-can-i-do-with-hooks-that-i-couldnt-with-classes}
```
Nie jestem pewien tego przecinka

> Hooki oferują nowy, potężny i wyrazisty sposób do wielokrotnego używania funkcjonalności pomiędzy komponentami. ["Tworzenie własnych hooków"](/docs/hooks-custom.html) zawiera szybki wgląd tego co jest możliwe. [Ten artykuł](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889) napisany przez jednego z głównym członków zespołu Reacta, zawiera bardziej szczegółowe informacje o nowych możliwościach otwartych przez hooki.

Reviewer 3: ```suggestion
Hooki oferują nowy, potężny i ekspresyjny sposób na wielokrotne używanie funkcjonalności pomiędzy komponentami. Rozdział p.t. ["Tworzenie własnych hooków"](/docs/hooks-custom.html) zawiera szybki wgląd tego co jest możliwe. [Ten artykuł](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889) napisany przez jednego z głównym członków zespołu Reacta, zawiera bardziej szczegółowe informacje o nowych możliwościach otwartych przez hooki.
```
Wydaje mi się, że _ekspresyjny_ lepiej oddaje to o czym mówimy dalej.

### Jaka część mojej wiedzy o Reakcie jest nadal aktualna? {#how-much-of-my-react-knowledge-stays-relevant}

> Hooki są bardziej bezpośrednim sposobem do użycia funkcjonalności React, które już znasz -- takie jak na przykład stan, metody cyklu życia, kontekst i referencje. Nie zmieniają podstaw działania Reakta, dlatego też twoja widza na temat komponentów, propsów i przepływu danych z góry w dół jest ciągle aktualna.

Reviewer 3: ```suggestion
Hooki są bardziej bezpośrednim sposobem do użycia funkcjonalności Reacta, które już znasz -- takie jak na przykład stan, metody cyklu życia (ang. *lifecycle methods*), kontekst i referencje (ang. *refs*). Nie zmieniają podstaw działania Reacta, dlatego też twoja wiedza na temat komponentów, właściwości (ang. *props*) i przepływu danych z góry w dół jest ciągle aktualna.
```

> Hooki są bardziej bezpośrednim sposobem do użycia funkcjonalności Reacta, które już znasz -- takie jak na przykład stan, metody cyklu życia (ang. *lifecycle methods*), kontekst i referencje (ang. *refs*). Nie zmieniają podstaw działania Reacta, dlatego też twoja wiedza na temat komponentów, właściwości (ang. *props*) i przepływu danych z góry w dół jest ciągle aktualna.

Reviewer 4: Wg. mnie tutaj "lifecycle" to po prostu "cykl życia komponentu", a nie "metody cyklu życia", które są specyficznym dla klas rozwiązaniem.

Ze względu na to, że hooki pojawiły się całkiem niedawno, niektóre biblioteki firm trzecich mogą być z nimi niekompatybilne.

### Czy hooki zastępują właściwości renderujące i komponenty wyższego rzędu? {#do-hooks-replace-render-props-and-higher-order-components}

Zazwyczaj właściwości renderujace i komponenty wyższego rzędu renderują tylko pojedynczy komponent potomny. Sądzimy, że hooki są prostszym sposobem na obsługę tego przypadku użycia. Nadal jest miejsce dla obu wzorców (dla przykładu, wirtualny komponent do obsługi suwaka może mieć właściwość `renderItem` lub widoczny komponent kontenera może mieć swoją własną strukturę DOM). Jednak w większości przypadków hooki w zupełności wystarczą, a przy okazji pomogą zmniejszyć liczbę zagnieżdżeń w drzewie.

### Co hooki oznaczają dla popularnych API takich jak Redux connect() i React Router?{#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

> Możesz używać tych samych API, co do tej pory; będą nadal działać.

Reviewer 3: Nie jestem pewien średnika

Author: @g12i tu się właśnie zastanawiałem nad myślnikiem lub przecinkiem, jak sądzisz co będzie odpowiedniejsze ? Lub może masz inny pomysł ?

Author: dałem myślink :)

> W przyszłości nowe wersje tych bibliotek mogą również eksportować spersonalizowane hooki, takie jak `useRedux()` czy `useRouter()`, które pozwolą na użycie tych samych funkcjonalności bez potrzeby opakowywania komponentów.

Reviewer 3: ```suggestion
W przyszłości nowe wersje tych bibliotek będą mogły również eksportować spersonalizowane hooki, takie jak `useRedux()` czy `useRouter()`, które pozwolą na użycie tych samych funkcjonalności bez potrzeby opakowywania komponentów.
```

### Czy hooki współpracują ze statycznym typowaniem? {#do-hooks-work-with-static-typing}

> Hooki zostały zaprojektowane z myślą o statycznym typowaniu. Dzięki temu że są funkcjami, łatwiej jest je poprawnie otypować w odróżnieniu od wzorców jak na przykład komponenty wyższego rzędu. Najnowsze definicje Reacta dla Flow i TypeScripta wspierają hooki.

Reviewer 3: ```suggestion
Hooki zostały zaprojektowane z myślą o statycznym typowaniu. Dzięki temu że są funkcjami, łatwiej jest je poprawnie otypować, w odróżnieniu od wzorców takich jak komponenty wyższego rzędu. Najnowsze definicje Reacta dla Flow i TypeScripta wspierają hooki.
```

Co ważne, spersonalizowane hooki dają możliwość ograniczenia API Reacta, jeżeli tylko zechcesz możesz je otypować bardziej ściśle. React udostępnia podstawy, ale możesz je łączyć na różne sposoby, odmienne od tych które dostarczyliśmy w standardzie.

### Jak testować komponenty które używają hooków? {#how-to-test-components-that-use-hooks}

> Z punktu widzenia Reacta, komponent który wykorzystuje hooki jest zwyczajnym komponentem. Jeżeli rozwiązanie które testujesz, nie opiera się na wnętrzu Reacta, to podejście dotyczące testowania komponentów, które używają hooków, nie powinno się różnić od tego jak robisz to zazwyczaj.

Reviewer 3: ```suggestion
Z punktu widzenia Reacta, komponent wykorzystujący hooki jest zwyczajnym komponentem. Jeżeli twoje rozwiązanie do testów nie opiera się na wewnętrznej implementacji Reacta, to testowanie komponentów, które używają hooków, nie powinno różnić się od tego co robisz to zazwyczaj.
```
Hmm, _rozwiązanie które testujesz_ byłoby _solution that you test_. _Your testing solution_ to _Rozwiązanie do testowania_. Ale przyznam szczerze, że nie rozumiem co autor miał na myśli.

### Co mogę zrobić z Hookami czego nie mogłem zrobić z klasami? {#what-can-i-do-with-hooks-that-i-couldnt-with-classes}

Hooki oferują nowy, potężny i wyrazisty sposób do wielokrotnego używania funkcjonalności pomiędzy komponentami. ["Tworzenie własnych hooków"](/docs/hooks-custom.html) zawiera szybki wgląd tego co jest możliwe. [Ten artykuł](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889) napisany przez jednego z głównym członków zespołu Reacta, zawiera bardziej szczegółowe informacje o nowych możliwościach otwartych przez Hooki.

### Jaka część mojej wiedzy o Reakcie jest nadal aktualna? {#how-much-of-my-react-knowledge-stays-relevant}

Hooki są bardziej bezpośrednim sposobem do użycia funkcjonalności React, które już znasz -- takie jak na przykład stan, metody cyklu życia, kontekst i referencje. Nie zmieniają podstaw działania Reakta, dlatego też twoja widza na temat komponentów, propsów i przepływu danych z góry w dół jest ciągle aktualna.

Hooki same w sobie, posiadają pewną krzywą uczenia się. Jeżeli brakuje czegoś w tej dokumentacji, [zgłoś problem](https://github.com/reactjs/reactjs.org/issues/new) a my postaramy się pomóc.

> ### Powinienem używać Hooków, klas, a może mieszać oba sposoby? {#should-i-use-hooks-classes-or-a-mix-of-both}

Reviewer 2: ```suggestion
### Czego mam używać: hooków, klas, a może mieszać oba sposoby? {#should-i-use-hooks-classes-or-a-mix-of-both}
```

> Jeżeli jesteś gotowy, zachęcamy do spróbowania Hooków w nowych komponentach, które piszesz. Upewnij się że wszyscy z twojego zespołu, wiedzą jak ich używać i są zapoznani z tą dokumentacją. Nie zalecamy przepisywania twoich istniejących klas do Hooków, chyba że z jakiegoś powodu i tak planowałeś je przepisać (na przykład aby naprawić błędy).

Reviewer 2: ```suggestion
Zachęcamy do wypróbowania hooków w nowych komponentach. Upewnij się, że wszyscy z twojego zespołu wiedzą, jak ich używać i są zapoznani z tą dokumentacją. Nie zalecamy przepisywania istniejących klas do hooków, chyba że z jakiegoś powodu i tak mieliście to w planach (na przykład, żeby naprawić błędy).
```

> Nie możesz używać Hooków *wewnątrz* komponentów klasowych, jednakże bez obaw możesz mieszać komponenty klasowe i funkcyjne z Hookami w pojedynczym drzewie. To czy komponent jest klasowy, czy funkcyjny i używa Hooków jest detalem implementacyjnym tego komponentu. W dłuższej perspektywie oczekujemy, że Hooki będą głównym sposobem pisania komponentów Reacta.

Reviewer 2: ```suggestion
Nie możesz używać Hooków *wewnątrz* komponentów klasowych, jednakże bez obaw możesz mieszać komponenty klasowe i funkcyjne z Hookami w tym samym drzewie. To, czy komponent jest klasowy, czy funkcyjny i używa Hooków jest detalem implementacyjnym tego komponentu. W dłuższej perspektywie oczekujemy, że hooki będą głównym sposobem pisania komponentów reactowych.
```

### Czy Hooki zapewniają wszystkie przypadki użycia, które są dostępne dla klas? {#do-hooks-cover-all-use-cases-for-classes}

> Naszym celem dla Hooków jest zapewnienie wszystkich przypadków użycia, które zapewniają klasy, tak szybko jak to tylko możliwe. Nie ma jeszcze odpowiednika w Hookach dla rzadkich cyklów życia komponentu takich jak `getSnapshotBeforeUpdate` i `componentDidCatch`, ale zamierzamy je wkrótce dodać.

Reviewer 2: ```suggestion
Naszym celem dla Hooków jest zapewnienie wszystkich przypadków użycia klas, tak szybko jak to tylko możliwe. Nie ma jeszcze odpowiednika w Hookach dla rzadziej używanych metod cyklu życia komponentu, takich jak `getSnapshotBeforeUpdate` i `componentDidCatch`, ale zamierzamy je wkrótce dodać.
```

> Ze względu na to, że jest to wczesny czas dla Hooków, niektóre biblioteki firm trzecich mogą być z nimi niekompatybilne.

Reviewer 2: ```suggestion
Ze względu na to, że hooki pojawiły się całkiem niedawno, niektóre biblioteki firm trzecich mogą być z nimi niekompatybilne.
```

### Czy Hooki zastępują właściwości renderujące i komponenty wyższego rzędu? {#do-hooks-replace-render-props-and-higher-order-components}

> Zazwyczaj, właściwości renderujace i komponenty wyższego rzędu renderują tylko pojedyncze dziecko. Sądzimy że Hooki są prostszym sposobem do obsługi tego przypadku użycia. Nadal jest miejsce dla obu wzorców (dla przykładu, wirtualny komponent do obsługi suwaka może mieć właściwość `renderItem` lub widoczny komponent kontenera może mieć swoją własną strukturę DOM). Jednak w większości przypadków, Hooki będą wystarczające i pomogą zmniejszyć zagnieżdżenia w drzewie.

Reviewer 2: ```suggestion
Zazwyczaj właściwości renderujace i komponenty wyższego rzędu renderują tylko pojedynczy komponent potomny. Sądzimy, że hooki są prostszym sposobem na obsługę tego przypadku użycia. Nadal jest miejsce dla obu wzorców (dla przykładu, wirtualny komponent do obsługi suwaka może mieć właściwość `renderItem` lub widoczny komponent kontenera może mieć swoją własną strukturę DOM). Jednak w większości przypadków hooki w zupełności wystarczą, a przy okazji pomogą zmniejszyć liczbę zagnieżdżeń w drzewie.
```

> ### Co Hooki oznaczają dla popularnych API takich jak Redux connect() i React Router?{#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

Reviewer 2: ```suggestion
### Co hooki oznaczają dla popularnych API takich jak connect() z Reduxa i React Router? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}
```

> Nadal możesz używać tych samych API, których używałeś do tej pory; będą nadal działać.

Reviewer 2: ```suggestion
Możesz używać tych samych API, co do tej pory; będą nadal działać.
```

> W przyszłości, nowe wersje tych bibliotek mogą również eksportować spersalizowane hooki, takie jak `useRedux()` czy `useRouter()`, które pozwolą na użycie tych samych funkcjonalności bez potrzeby opakowywania komponentów.

Reviewer 2: ```suggestion
W przyszłości nowe wersje tych bibliotek mogą również eksportować spersonalizowane hooki, takie jak `useRedux()` czy `useRouter()`, które pozwolą na użycie tych samych funkcjonalności bez potrzeby opakowywania komponentów.
```

>   * [Co mogę zrobić z Hookami czego nie mogłem zrobić z klasami?](#what-can-i-do-with-hooks-that-i-couldnt-with-classes)

Reviewer 2: Uzgodniliśmy, że 'hooki' piszemy z małej litery.

  * [Jaka część mojej wiedzy o Reakcie jest nadal aktualna?](#how-much-of-my-react-knowledge-stays-relevant)
  * [Powinienem używać Hooków, klas, a może mieszać oba sposoby?](#should-i-use-hooks-classes-or-a-mix-of-both)

>   * [Czy Hooki zapewniają wszystkie przypadki użycia, które są dostępne dla klas?](#do-hooks-cover-all-use-cases-for-classes)

Reviewer 2: Z tego co mi wiadomo, zwykle się "pokrywa przypadki użycia", a nie "zapewnia". Można też "zapewnić pokrycie przypadków użycia" :-)

Author: Masz rację, wcześniej było poprawnie, zmieniłem znów na "pokrywa". Nie pamiętam już dlaczego dałem "zapewnia".

> Jeżeli jesteś gotowy, zachęcamy do spróbowania Hooków w nowych komponentach, które piszesz. Upewnij się że wszyscy z twojego zespołu, wiedzą jak ich używać i są zapoznani z tą dokumentacją. Nie zalecamy przepisywania twoich istniejących klas do Hooków, chyba że z jakiegoś powodu i tak planowałeś je przepisać (na przykład aby naprawić błędy).

Reviewer 2: Zachęcamy do wypróbowania hooków w nowych komponentach. Upewnij się, że wszyscy z twojego zespołu wiedzą, jak ich używać i są zapoznani z tą dokumentacją. Nie zalecamy przepisywania istniejących klas do hooków, chyba że z jakiegoś powodu i tak mieliście to w planach (na przykład, żeby naprawić błędy).

  container = null;
});

it('potrafi wyrenderować i aktualizować licznik', () => {
  // Testuje pierwsze renderowanie i efekt
  act(() => {
    ReactDOM.render(<Counter />, container);
  });
  const button = container.querySelector('button');
  const label = container.querySelector('p');
  expect(label.textContent).toBe('Kliknięto 0 razy');
  expect(document.title).toBe('Kliknięto 0 razy');

  // Testuje drugie renderowanie i efekt
  act(() => {
    button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  expect(label.textContent).toBe('Kliknięto 1 razy');
  expect(document.title).toBe('Kliknięto 1 razy');
});
```

> Wywołanie `act()` spowoduje również znajdujących się w ich wnętrzu efektów.

Reviewer 3: ```suggestion
Wywołanie funkcji `act()` opróżni bufor efektów znajdujących się wewnątrz.
```
Tu również nie jestem pewien co autor miał na myśli, pisząc _inside of them_.

> Jeżeli musisz przetestować spersonalizowany hook, możesz to zrobić tworząc komponent w swoim teście i używając z niego hooka. Następnie, możesz napisać test do stworzonego w ten sposób komponentu.

Reviewer 3: ```suggestion
Jeżeli musisz przetestować spersonalizowany hook, możesz stworzyć komponent w swoim teście i wywołać ten hook w ciele jego funkcji. Następnie, możesz napisać test do stworzonego w ten sposób komponentu.
```

Aby zmniejszyć powtarzalność kodu, zalecamy użyć [`react-testing-library`](https://git.io/react-testing-library), która została zaprojektowana, aby zachęcać do pisania testów, które używają komponentów, w taki sam sposób jak robią to użytkownicy końcowi.

### Czego dokładnie wymagają [reguły lintera](https://www.npmjs.com/package/eslint-plugin-react-hooks)? {#what-exactly-do-the-lint-rules-enforce}

> Zapewniamy [wtyczkę do ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks) która, aby uniknąć błędów, zmusza do przestrzegania [zasad hooków](/docs/hooks-rules.html). Zakłada ona, że każda funkcja zaczynająca się od "`use`" i zaraz po tym, dużej litery, jest hookiem. Rozpoznanie tej heurystyki, nie jest idealne i może wystąpić wiele fałszywych alarmów, ale bez konwencji dla całego ekosystemu, nie ma możliwości aby hooki działały poprawnie -- dłuższe nazwy zniechęcą ludzi do używania hooków lub do przestrzegania konwencji.

Reviewer 3: ```suggestion
Stworzyliśmy [wtyczkę do ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks) która, aby uniknąć błędów, zmusza do przestrzegania [zasad hooków](/docs/hooks-rules.html). Zakłada ona, że każda funkcja zaczynająca się od "`use`" i zaraz po tym wielkiej litery, jest hookiem. Zdajemy sobie sprawę, że ta heurystyka nie jest idealna i może wywołać wiele fałszywych alarmów. Ale bez wprowadzenia wspólnej dla całego ekosystemu konwencji, nie ma możliwości aby hooki działały poprawnie -- dłuższe nazwy zniechęcą ludzi do używania hooków lub do przestrzegania tej konwencji.
```

> W szczegółach, zasada ta, wymusza aby:

Reviewer 3: ```suggestion
W szczególności zasada ta wymusza aby:
```

> * Wywołania hooków znajdowały się wewnątrz funkcji pisanej stylem `PascalCase` (zakłada że jest to komponent) lub innej funkcji `useSomething` (zakłada że jest to spersonaliowany hook).

Reviewer 3: ```suggestion
* Wywołania hooków znajdowały się wewnątrz funkcji pisanej stylem `PascalCase` (zakłada że jest to komponent) lub innej funkcji `useSomething` (zakłada że jest to własny hook).
```

* Hooki są wywoływane w tej samej kolejnosći, przy każdym renderze.

> Jest jeszcze kilka heurystyk i mogą się one zmieniać wraz z czasem, gdy dostroimy zasadę tak, aby zbalansować wyszukiwanie błędów i zmniejszyć dawanie fałszywych alarmów.

Reviewer 3: ```suggestion
Jest jeszcze kilka innych heurystyk i mogą się one z czasem zmienić, gdy dostroimy zasadę tak, aby zbalansować wyszukiwanie błędów i zmniejszyć wywoływanie fałszywych alarmów.
```

}
```

Zauważ, jak mogliśmy przenieść wywołanie `useState` dla zmiennej stanu `position` i powiązany z nią efekt do własnego hooka, bez konieczności zmiany jego kodu. Jeżeli cały stan znajdowałby się w pojedynczym obiekcie, wyodrębnienie go byłoby trudniejsze.

Zarówno umieszczanie całego stanu wewnątrz pojedynczego wywołania `useState`, jak i wywoływanie `useState` dla każdego pola, będzie działać. Komponenty będą najbardziej czytelne, jeżeli odnajdziesz równowagę pomiędzy tymi dwoma skrajnościami i pogrupujesz powiązane ze sobą zmienne stany. Jeżeli logika stanu stanie się zbyt złożona, zalecamy [użycie reduktora](/docs/hooks-reference.html#usereducer) lub napisanie własnego hooka.

> ### Czy mogę uruchomić efekt tylko podczas aktualizacji? {#can-i-run-an-effect-only-on-updates}

Reviewer 2: ```suggestion
### Czy mogę uruchomić efekt tylko podczas aktualizacji komponentu? {#can-i-run-an-effect-only-on-updates}
```

  });
  const prevCount = prevCountRef.current;

  return <h1>Teraz: {count}, poprzednio: {prevCount}</h1>;
}
```

> Może to być trochę zawiłe, ale możesz to wyodrębnić do własnego hooka:

Reviewer 2: ```suggestion
Może to wydawać się trochę zawiłe, ale wystarczy wyodrębnić tę logikę do osobnego hooka:
```

}
```

> Zauważ że, będzie to działać zarówno dla właściwości, stanu i każdej innej wyliczanej wartości.

Reviewer 2: ```suggestion
Zauważ, że powyższa funkcja zadziała poprawnie dla właściwości, zmiennej stanu oraz każdej innej wyliczanej wartości.
```

  // ...
```

> Ponieważ jest to powszechny przypadek użycia, bardzo prawdopodobne że w przyszłości, React sam dostarczy implementacje hooka `usePrevious`.

Reviewer 2: ```suggestion
Ponieważ jest to powszechny przypadek użycia, bardzo prawdopodobne, że w przyszłości React sam będzie dostarczał implementację hooka `usePrevious`.
```

Spójrz również na [rekomendowany wzorzec dla stanu pochodnego](#how-do-i-implement-getderivedstatefromprops).

### Dlaczego widzę nieaktualne propsy lub stan wewnątrz mojej funkcji? {#why-am-i-seeing-stale-props-or-state-inside-my-function}

> Każda funkcja wewnątrz komponentu, włączając w to uchwyty zdarzeń i efekty, "widzą" propsy i stan, aktualny na czas rendera, w którym zostały stworzone. Dla przykładu, rozważ poniższy kod:

Reviewer 2: ```suggestion
Każda funkcja wewnątrz komponentu, włączając w to procedury obsługi zdarzeń i efekty, "widzą" właściwości i stan z chwili renderowania, w którym zostały stworzone. Dla przykładu rozważ poniższy kod:
```

```js
function Example() {
  const [count, setCount] = useState(0);

  function handleAlertClick() {
    setTimeout(() => {
      alert('Kliknięto: ' + count);
    }, 3000);
  }

  return (
    <div>
      <p>Kliknięto {count} razy</p>
      <button onClick={() => setCount(count + 1)}>
        Naciśnij mnie
      </button>
      <button onClick={handleAlertClick}>
        Pokaż okno ostrzegawcze
      </button>
    </div>
  );
}
```

> Jeżeli najpierw klikniesz "Pokaż okno ostrzegawcze", a następnie zinkrementujesz licznik, okno ostrzegawcze wyświetli watość zmiennej `count` **z momentu kliknięcia na przycisk "Pokaż okno ostrzegawcze"**. Zapobiega to błędom w kodzie, który zakłada że propsy i stan nie zmienią się.

Reviewer 2: ```suggestion
Jeżeli najpierw klikniesz "Pokaż okno ostrzegawcze", a następnie zwiększysz licznik, okno ostrzegawcze wyświetli wartość zmiennej `count` **z momentu kliknięcia na przycisk "Pokaż okno ostrzegawcze"**. Zapobiega to błędom powodowanym przez kod zakładający, że właściwości i stan nie zmienią się w czasie.
```

> Jeżeli celowo chcesz odczytać *najświeższy* stan z asynchronicznej pętli zwrotnej, możesz go przechowywać, zmieniać i odczytywać korzystając z [referencji](/docs/hooks-faq.html#is-there-something-like-instance-variables).

Reviewer 2: ```suggestion
Jeżeli celowo chcesz odczytać *najświeższy* stan z wnętrza asynchronicznej funkcji zwrotnej, możesz go przechowywać, zmieniać i odczytywać korzystając z [referencji](/docs/hooks-faq.html#is-there-something-like-instance-variables).
```

> Ostatecznie, inną możliwą przyczyną tego że widzisz nieaktualne propsy lub stan, może być użycie "tablicy zależności" do optymilizacji, ale niepoprawne sprecyzowanie wszystkich zależności. Dla przykładu, jeżeli efekt otrzymuje `[]` jako drugi argument, ale wewnątrz odczytuje `someProp`, efekt będzie stale "widział" początkową wartość `someProp`. Rozwiązaniem jest usunięcie tablicy zależności lub naprawienie jej. Tutaj znajdziesz informacje [jak poradzić sobie z funkcjami](#is-it-safe-to-omit-functions-from-the-list-of-dependencies), a tutaj [inne powszechne sposoby](#what-can-i-do-if-my-effect-dependencies-change-too-often), aby uruchamiać efekt rzadziej i bez niepoprawnego pomijania zależności.

Reviewer 2: ```suggestion
Ostatecznie, inną możliwą przyczyną tego, że widzisz nieaktualne właściwości lub stan, może być użycie "tablicy zależności" do optymalizacji, ale niepoprawne sprecyzowanie wszystkich zależności. Dla przykładu, jeżeli efekt otrzymuje `[]` jako drugi argument, ale wewnątrz odczytuje `someProp`, efekt będzie stale "widział" początkową wartość `someProp`. Rozwiązaniem jest usunięcie tablicy zależności lub naprawienie jej. Tutaj znajdziesz informacje, [jak poradzić sobie z funkcjami](#is-it-safe-to-omit-functions-from-the-list-of-dependencies), a tutaj [inne powszechne sposoby](#what-can-i-do-if-my-effect-dependencies-change-too-often) na uruchamianie efektów rzadziej i bez błędów w zależnościach.
```

>Uwaga
>

> >Stworzyliśmy regułę [`wyczerpującą zależności`](https://github.com/facebook/react/issues/14920) dla ESLint, jako część paczki [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation). Wtyczka ostrzega, gdy zależności są sprecyzowane niepoprawnie i zaleca poprawienie kodu.

Reviewer 2: ```suggestion
>Stworzyliśmy regułę [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) dla ESLinta i dodaliśmy ją do paczki [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation). Wtyczka ostrzega, gdy zależności są sprecyzowane niepoprawnie i zaleca poprawienie kodu.
```

### Jak zaimplementować `getDerivedStateFromProps`? {#how-do-i-implement-getderivedstatefromprops}

> Prawdopodobnie [tego nie potrzebujesz](/blog/2018/06/07/you-probably-dont-need-derived-state.html), w rzadkich przypadkach w których naprawdę będziesz tego potrzebować (na przykład implementacja komponentu `<Transition>`), możesz zaktualizować stan w trakcie renderowania. React uruchomi ponownie komponent z zaktualizowanym stanem natychmiast po pierwszym renderowaniu.

Reviewer 2: ```suggestion
Prawdopodobnie [tego w ogóle nie potrzebujesz](/blog/2018/06/07/you-probably-dont-need-derived-state.html). W rzadkich przypadkach, w których naprawdę będziesz tego potrzebować (na przykład implementacja komponentu `<Transition>`), możesz zaktualizować stan w trakcie renderowania. React uruchomi ponownie komponent z zaktualizowanym stanem natychmiast po pierwszym renderowaniu, więc nie wpłynie to znacząco na wydajność.
```

> Poniżej, przechowujemy poprzednią wartość propsa `row` w zmiennej stanowej, więc możemy wykonać porównanie:

Reviewer 2: ```suggestion
W poniższym kodzie przechowujemy poprzednią wartość właściwości `row` w zmiennej stanowej, dzięki czemu możemy wykonać porównanie:
```

```js
function ScrollView({row}) {
  let [isScrollingDown, setIsScrollingDown] = useState(false);
  let [prevRow, setPrevRow] = useState(null);

  if (row !== prevRow) {

>     // Wiersz się zmienił od ostatniego renderowania. Zaktualizuj isScrollingDown.

Reviewer 2: ```suggestion
    // Wiersz zmienił się od ostatniego renderowania. Zaktualizuj isScrollingDown.
```

    setIsScrollingDown(prevRow !== null && row > prevRow);
    setPrevRow(row);
  }

  return `Przewijanie w dół: ${isScrollingDown}`;
}
```

> Na pierwszy rzut oka może to wyglądać dziwnie, ale aktualizacja podczas renderowania jest dokładnie tym samym, czym w założeniach  `getDerivedStateFromProps` był od zawsze.

Reviewer 2: ```suggestion
Na pierwszy rzut oka może to wyglądać dziwnie, ale aktualizacja podczas renderowania jest dokładnie tym samym, czym w założeniu metoda `getDerivedStateFromProps` była od zawsze.
```

### Czy istnieje coś takiego jak forceUpdate? {#is-there-something-like-forceupdate}

> Zarówno `useState` jak i `useReducer` [wycofują się z aktualizacji](/docs/hooks-reference.html#bailing-out-of-a-state-update), jeżeli kolejna wartość jest taka sama jak poprzednia. Zmiana stanu bez użycia `setState`, a następnie wywołanie `setState` nie spowoduje przerenderowania.

Reviewer 2: ```suggestion
Zarówno `useState`, jak i `useReducer` [wycofują się z aktualizacji](/docs/hooks-reference.html#bailing-out-of-a-state-update), jeżeli kolejna wartość jest taka sama jak poprzednia. Zmiana stanu bez użycia `setState`, a następnie wywołanie `setState` nie skutkuje ponownym renderowaniem komponentu.
```

> Zazwyczaj, nie powinno się modyfikować lokalnego stanu Reacta. Jednakże, możesz użyć inkrementacji licznika, aby wymusić przerenderowanie, nawet gdy stan się nie zmienił:

Reviewer 2: ```suggestion
Zazwyczaj nie powinno się modyfikować lokalnego stanu w Reakcie. Możesz jednak inkrementować licznik, aby wymusić ponowne renderowanie, nawet jeśli stan się nie zmienił:
```

  }
```

Jeżeli to możliwe, staraj się unikać tego wzorca.

### Czy mogę stworzyć referencję do komponentu funkcyjnego? {#can-i-make-a-ref-to-a-function-component}

> Nie powinno się tego robić zbyt często, jednak możesz odsłonić niektóre imperatywne metody dla komponentu rodzica używając hooka [`useImperativeHandle`](/docs/hooks-reference.html#useimperativehandle).

Reviewer 2: ```suggestion
Nie powinno się tego robić zbyt często, jednak możesz upublicznić niektóre imperatywne metody dla komponentu rodzica używając hooka [`useImperativeHandle`](/docs/hooks-reference.html#useimperativehandle).
```

### Jak mogę zmierzyć węzeł DOM? {#how-can-i-measure-a-dom-node}

> Aby zmierzyć pozycję lub rozmiar węzła DOM, możesz użyć [referencji pętli zwrotnej](/docs/refs-and-the-dom.html#callback-refs). React wywoła pętle zwrotną gdy ref zostanie zaczepiony do innego węzła. Tutaj znajdziesz [przykładowe demo](https://codesandbox.io/s/l7m0v5x4v9):

Reviewer 2: ```suggestion
Aby zmierzyć pozycję lub rozmiar węzła DOM, możesz użyć [referencji z funkcją zwrotną](/docs/refs-and-the-dom.html#callback-refs). React wywoła funkcję zwrotną, gdy referencja zostanie przypisana do innego węzła. Tutaj znajdziesz [prosty przykład](https://codesandbox.io/s/l7m0v5x4v9):
```

  return (
    <>

>       <h1 ref={measuredRef}>Witaj świecie</h1>

Reviewer 2: ```suggestion
      <h1 ref={measuredRef}>Witaj, świecie</h1>
```

>       <h2>Powyższy nagłówek ma {Math.round(height)}pikseli wysokości</h2>

Reviewer 2: ```suggestion
      <h2>Powyższy nagłówek ma {Math.round(height)} pikseli wysokości</h2>
```

    </>
  );
}
```

> W tym przykładzie nie zdecydowaliśmy się użyć `useRef`, ponieważ obiekt referencji nie powiadamia nas o *zmiannach* jego aktualnej wartości. Użycie referencji pętli zwrotnej działa [nawet gdy komponent dziecka wyświetla mierzony węzeł później](https://codesandbox.io/s/818zzk8m78) (np. w odpowiedzi na przyciśnięcie przycisku), o tym zdarzeniu informowany jest komponent rodzica, który może zaktualizować pomiary.

Reviewer 2: ```suggestion
W tym przykładzie nie zdecydowaliśmy się użyć `useRef`, ponieważ obiekt referencji nie powiadamia nas o *zmianach* jego aktualnej wartości. Użycie referencji z funkcją zwrotną daje pewność, że [nawet jeśli komponent potomny wyświetli mierzony węzeł później](https://codesandbox.io/s/818zzk8m78) (np. w odpowiedzi na przyciśnięcie przycisku), komponent nadrzędny zostanie o tym powiadomiony i może zaktualizować swój pomiar.
```

> Zuważ że przekazaliśmy `[]`, jako tablice zależności do `useCallback`. Gwartantuje to nam niezmienialność pętli zwrotnej pomiedzy re-renderami, oraz że React nie wywoła jej bez potrzeby.

Reviewer 2: ```suggestion
Zauważ, że przekazaliśmy `[]` jako tablicę zależności do `useCallback`. Gwarantuje to nam niezmienialność funkcji zwrotnej pomiedzy ponownymi renderowaniami oraz że React nie wywoła jej bez potrzeby.
```

> W razie potrzeby, można [wyodrębnić tę logikę](https://codesandbox.io/s/m5o42082xy) do wielokrotnego użycia w hooku:

Reviewer 2: ```suggestion
W razie potrzeby można [wyodrębnić tę logikę](https://codesandbox.io/s/m5o42082xy) do osobnego hooka i używać wielokrotnie:
```

```js{2}
function MeasureExample() {
  const [rect, ref] = useClientRect();
  return (
    <>

>       <h1 ref={ref}>Witaj świecie</h1>

Reviewer 2: ```suggestion
      <h1 ref={ref}>Witaj, świecie</h1>
```

      {rect !== null &&

>         <h2>Powyższy nagłówek ma {Math.round(rect.height)}pikseli wysokości</h2>

Reviewer 2: ```suggestion
        <h2>Powyższy nagłówek ma {Math.round(rect.height)} pikseli wysokości</h2>
```
